provider/views.py

```diff
@@ -262,11 +265,48 @@ def handle(self, request, post_data=None):
         authorization_form = self.get_authorization_form(request, client,
             post_data, data)
 
-        if not authorization_form.is_bound or not authorization_form.is_valid():
-            return self.render_to_response({
-                'client': client,
-                'form': authorization_form,
-                'oauth_data': data, })
+        if data.get('response_type', None):
+            if not (authorization_form.is_bound and authorization_form.is_valid()):
+                return self.render_to_response({
+                    'client': client,
+                    'form': authorization_form,
+                    'oauth_data': data,
+                })
+
+            if not authorization_form.cleaned_data['authorize']:
+                # User hit cancel.
+                response_data = {
+                    "error": "access_denied",
+                    "error_description": "The user denied the scope request.",
+                    "state": data.get('state'),
+                }
+
+                url = "{0}?{1}".format(
+                    client.redirect_uri, urllib.urlencode(response_data))
+
+                return HttpResponseRedirect(url)
+
+            if data.get('response_type') == 'token':
+                #uses request.user as the urls.py already required login_required
+                create_kwargs = {
+                    "user": request.user,
+                    "client": client,
+                }
+                if data.get('scope'):
+                    create_kwargs['scope'] = data.get('scope')
+
+                atm = None
+                if constants.SINGLE_ACCESS_TOKEN:
+                    try:
+                        atm = AccessTokenModel.objects.get(user=request.user)
```

Comment:
oops, master has a bug that this branch does not
https://github.com/justinabrahms/django-oauth2-provider/blob/grant_type_token/provider/oauth2/views.py#L106

the expiration date has to be checked as well.